sklearn/linear_model/tests/test_sag.py

```diff
@@ -0,0 +1,710 @@
+                intercept -= eta * (intercept_sum_gradient / len(seen)) * decay
+
+            gradient_memory[idx] = gradient
+
+            wscale *= (1.0 - alpha * eta)
+            if counter == 0:
+                c_sum[0] = eta / (wscale * len(seen))
+            else:
+                c_sum[counter] = (c_sum[counter - 1] +
+                                  eta / (wscale * len(seen)))
+
+            if counter >= 1 and wscale < 1e-9:
+                for j in range(n_features):
+                    if last_updated[j] == 0:
+                        weights[j] -= c_sum[counter] * sum_gradient[j]
+                    else:
+                        weights[j] -= ((c_sum[counter] -
+                                        c_sum[last_updated[j] - 1]) *
+                                       sum_gradient[j])
+                    last_updated[j] = counter + 1
+                c_sum[counter] = 0
+                weights *= wscale
+                wscale = 1.0
+
+            counter += 1
+
+    for j in range(n_features):
+        if last_updated[j] == 0:
+            weights[j] -= c_sum[counter - 1] * sum_gradient[j]
+        else:
+            weights[j] -= ((c_sum[counter - 1] -
+                            c_sum[last_updated[j] - 1]) *
+                           sum_gradient[j])
+    weights *= wscale
+    return weights, intercept
+
+
+def get_eta(X, alpha, fit_intercept, classification=True):
+    if classification:
+        return (4.0 / (np.max(np.sum(X * X, axis=1))
+                + fit_intercept + 4.0 * alpha))
+    else:
+        return 1.0 / (np.max(np.sum(X * X, axis=1)) + fit_intercept + alpha)
+
+
+@ignore_warnings
+def test_classifier_matching():
+    n_samples = 20
+    X, y = make_blobs(n_samples=n_samples, centers=2, random_state=0,
+                      cluster_std=0.1)
+    y[y == 0] = -1
+    alpha = 1.1
+    n_iter = 80
+    fit_intercept = True
+    eta = get_eta(X, alpha, fit_intercept)
+    clf = LogisticRegression(solver="sag", fit_intercept=fit_intercept,
+                             tol=.00000000001, C=1. / alpha / n_samples,
+                             max_iter=n_iter, random_state=10)
+    clf.fit(X, y)
+
+    weights, intercept = sag_sparse(X, y, eta, alpha, n_iter=n_iter,
+                                    dloss=log_dloss,
+                                    fit_intercept=fit_intercept)
+    weights2, intercept2 = sag(X, y, eta, alpha, n_iter=n_iter,
+                               dloss=log_dloss,
+                               fit_intercept=fit_intercept)
+    weights = np.atleast_2d(weights)
+    intercept = np.atleast_1d(intercept)
+    weights2 = np.atleast_2d(weights2)
+    intercept2 = np.atleast_1d(intercept2)
+
+    assert_array_almost_equal(weights, clf.coef_, decimal=10)
+    assert_array_almost_equal(intercept, clf.intercept_, decimal=10)
+    assert_array_almost_equal(weights2, clf.coef_, decimal=10)
+    assert_array_almost_equal(intercept2, clf.intercept_, decimal=10)
```

Comment:
I agree it might have been useful as a scaffold when initially implementing and debugging the Cython version but now that we have other convergence tests to detect regression I am not sure we should keep the 2 Python variants in the tests from a maintenance point of view.

What do other people think (@GaelVaroquaux @agramfort @TomDLT ) ?